optimade/adapters/structures/pymatgen.py

```diff
@@ -0,0 +1,105 @@
+from typing import Union, Dict, List
+
+from optimade.models import Species as OptimadeStructureSpecies
+from optimade.models import StructureResource as OptimadeStructure
+from optimade.models.structures import Periodicity
+from optimade.adapters.exceptions import ConversionError
+
+try:
+    from pymatgen import Structure, Molecule
+
+except (ImportError, ModuleNotFoundError):
+    from warnings import warn
+
+    Structure = None
+    Molecule = None
+    PYMATGEN_NOT_FOUND = "Pymatgen not found, cannot convert structure to a pymatgen Structure or Molecule"
+
+
+__all__ = ("get_pymatgen",)
+
+
+def get_pymatgen(optimade_structure: OptimadeStructure) -> Union[Structure, Molecule]:
+    """ Get pymatgen Structure or Molecule from OPTIMADE structure
+
+    :param optimade_structure: OPTIMADE structure
+    :return: pymatgen.Structure , pymatgen.Molecule
+    """
+    if globals().get("Structure", None) is None:
+        warn(PYMATGEN_NOT_FOUND)
+        return None
+
+    # pymatgen cannot handle unknown positions
+    for site in optimade_structure.attributes.cartesian_site_positions:
+        if None in site:
+            raise ConversionError(
+                "pymatgen cannot be used to convert structures with unknown positions."
+            )
+
+    if optimade_structure.attributes.dimension_types == (Periodicity.PERIODIC,) * 3:
+        return _get_structure(optimade_structure)
+
+    return _get_molecule(optimade_structure)
+
+
+def _get_structure(optimade_structure: OptimadeStructure) -> Structure:
+    """Create pymatgen Structure from OPTIMADE structure"""
+
+    attributes = optimade_structure.attributes
+
+    return Structure(
+        lattice=attributes.lattice_vectors,
```

Comment:
Pymatgen doesn't return a `pymatgen.Structure`, but rather a `pymatgen.Molecule` for OPTIMADE structures with unknown lattice vectors (assuming the `dimension_types` are set accordingly). `pymatgen.Molecule` doesn't use `lattice_vectors` for whatever reason, I guess since it doesn't actually define a structure.
If you know more about pymatgen, you're welcome to correct it, so that we don't unneccesarily lose information during the conversion that we could've kept.